### DISCUSSION TITLE
Accessing the full URI in middleware

### DISCUSSION THREAD

**Discussion by rouge8** [tokio-rs/axum#1149] [Q&A] [answered]
I'm trying to write a middleware that redirects from HTTP -> HTTPS, and to do that, I need to be able to access the full URI in the middleware. I saw in #858 that this might be possible by extracting `axum::http::Uri`, but that doesn't appear to be available off of `Request.extensions()`.

```rust
/// Redirect HTTP requests to HTTPS
pub struct HTTPSRedirectLayer {
    enabled: bool,
}

impl HTTPSRedirectLayer {
    pub fn new(enabled: bool) -> Self {
        Self { enabled }
    }
}

impl<S> Layer<S> for HTTPSRedirectLayer {
    type Service = HTTPSRedirectMiddleware<S>;

    fn layer(&self, inner: S) -> Self::Service {
        HTTPSRedirectMiddleware {
            enabled: self.enabled,
            inner,
        }
    }
}

#[derive(Clone)]
pub struct HTTPSRedirectMiddleware<S> {
    enabled: bool,
    inner: S,
}

impl<S> Service<Request<Body>> for HTTPSRedirectMiddleware<S>
where
    S: Service<Request<Body>, Response = Response> + Send + 'static,
    S::Future: Send + 'static,
{
    type Response = S::Response;
    type Error = S::Error;
    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;

    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        self.inner.poll_ready(cx)
    }

    fn call(&mut self, req: Request<Body>) -> Self::Future {
        let uri = req.extensions().get::<Uri>().unwrap();

        dbg!(self.enabled, &uri, &req);

        let future = self.inner.call(req);
        Box::pin(async move {
            let response: Response = future.await?;
            Ok(response)
        })
    }
}
```

**Comment by davidpdrsn** [comment, marked as answer]
The uri is available directly in the request https://docs.rs/http/0.2.8/http/request/struct.Request.html#method.uri. No extensions needed.

**Reply by rouge8** [reply]
Hmm, `req.uri()` is only the path, not the full URI 😕

**Reply by davidpdrsn** [reply]
Depending on the scheme you might need to combine it with `Host`

**Reply by rouge8** [reply]
Won't that only give me the hostname, not the scheme?

**Reply by davidpdrsn** [reply]
The scheme should be available on the uri

**Reply by rouge8** [reply]
Oh! I looked at the docs and saw `assert_eq!(*request.uri(), *"/");` and assumed the uri was a string, not a struct. Thanks!

**Comment by ITmind** [comment]
Scheme not available on the uri. How to define it?
I use extractor OriginalUri and Host and "let url = format!("{}{}", hostname, uri.to_string())", but result is:
"127.0.0.1:3000/myreq", not "http://127.0.0.1:3000/myreq"

**Reply by jplatte** [reply]
This really depends, because at the http level, this information is not sent to the server. How do you run your axum server such that it allows more schemes than plain HTTP? Is there a reverse proxy in front? Are you using axum-server or some other TLS connector within your Rust app?
